src/cascade/model/age_time_grid.py

```diff
@@ -0,0 +1,153 @@
+from datetime import timedelta
+from itertools import product
+from math import nan, inf
+
+import numpy as np
+import pandas as pd
+
+from cascade.dismod.constants import PriorKindEnum
+
+GRID_SNAP_DISTANCE = 1 / timedelta(days=365).total_seconds()
+"""Times within one second are considered equal."""
+
+
+class AgeTimeGrid:
+    """The AgeTime grid holds rows of a table at each age and time value.
+
+    At each age and time point is a dataframe consisting of the columns
+    given in the constructor. So getting an item returns a dataframe
+    with those columns. Setting a dataframe sets those columns.
+    Each AgeTimeGrid has three possible mulstds, for value, dage, dtime.
+
+    >>> atg = AgeTimeGrid(([0, 10, 20], [1990, 2000, 2010]), ["height", "weight"])
+    >>> atg[:, :] = [6.1, 195]
+    >>> atg[:, :].height = [5.9]
+    >>> atg[10, 2000] = [5.7, 180]
+    >>> atg[5:17, 1980:1990].weight = 125
+    >>> assert (atg[20, 2000].weight == 195).all()
+    >>> assert isinstance(atg[0, 1990], pd.DataFrame)
+
+    If the column has the same name as a function (mean), then access it
+    with getitem,
+
+    >>> atg[:, :]["mean"] = [5.9]
+
+    """
+    def __init__(self, age_time_grid, columns, count=1):
+        assert isinstance(columns[0], str)
+        try:
+            self.ages = np.array(age_time_grid[0], dtype=np.float)
+            self.times = np.array(age_time_grid[1], dtype=np.float)
+        except TypeError:
+            raise TypeError(f"Ages and times should be arrays of floats {age_time_grid}.")
+        if isinstance(columns, str):
+            columns = [columns]
+        try:
+            self.columns = list(columns)
+        except TypeError:
+            raise TypeError(f"Columns should be an iterable of strings. {columns}")
+        for col_is_str in self.columns:
+            if not isinstance(col_is_str, str):
+                raise TypeError(f"Columns should be iterable of strings. {col_is_str}")
+        try:
+            count = int(count)
+        except ValueError:
+            raise TypeError(f"Count must be an integer {count}.")
+
+        age_time = np.array(list(product(sorted(self.ages), sorted(self.times))))
+        self.grid = pd.DataFrame(dict(
+            age=np.tile(age_time[:, 0], count),
+            time=np.tile(age_time[:, 1], count),
+            idx=np.repeat(range(count), len(age_time)),
+        ))
+        self.grid = self.grid.assign(**{new_col: nan for new_col in columns})
+        self._mulstd = dict()
+        # Each mulstd is one record.
+        for kind in PriorKindEnum:
+            mulstd_df = pd.DataFrame(dict(
+                age=[nan],
+                time=[nan],
+                idx=list(range(count)),
+            ))
+            mulstd_df = mulstd_df.assign(**{new_col: nan for new_col in columns})
+            self._mulstd[kind.name] = mulstd_df
+
+    @property
+    def mulstd(self):
+        return self._mulstd
```

Comment:
Is the intention that this be imutable? Might want to return a copy.

Reply:
What's in the dict has to be mutable. I just didn't want people assigning to the mulstd property... too much?

Reply:
No, that's fine. It's just that I usually use this pattern to produce immutability, both superficial and deep, but if that's not what you want then this is fine.